### ISSUE TITLE
Crash course in DosBox-X architecture and how things are "dispatched"

### ISSUE THREAD

**Issue by ThomasKjoernes** [joncampbell123/dosbox-x#4172]
### Question

Is there a description of how DosBox-X is glued together and how the various components works in relation to eachother?

Example:

1) If I wanted to add a custom hardware device. 
- How would a register it and make it respond to I/O and memory access
- Either by manually assigned address or how to make it a PnP (PCI) device etc.

2) If I want to implement realmode interrupt handlers, how is this registerred?
- Can you inject custom code/handlers into the DOS or BIOS?
- I assume handlers have "stubs" inside conventional memory that simply does nothing except tell DosBox-X to call a callback handler?

3) Can I intercept and implement unknown/custom cpu opcodes?

4) If I wanted to implement a custom device with a processor on it, how would I make this execute it's logic. 
- Do I create a separate thread or register timer handlers
- Or just respond to I/O-access and run code on the fly?

### Have you checked that no similar question(s) exist?

- [X] I have searched and didn't find any similar question.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by joncampbell123** [comment]
There is the DOSBox way (SVN, and all other forks) in which you declare an IO_ReadHandleObject and IO_WriteHandleObject and call the Install() method that specifies the I/O port, callback function on I/O access, and memory size accesses that trigger it (byte, word, dword) which itself is a bitmask.

Once registered, access to that I/O port will call your function which acts on the I/O. If a read callback, the return value is given to the guest. You might look at the Sound Blaster code as an example.

For memory, the DOSBox way is to assign a callback to one page (4KB) of memory using MEM_SetPageHandler(). The address is specified in 4KB pages, not bytes. Look at src/hardware/vga_memory.cpp for a good example.

**Comment by joncampbell123** [comment]
DOSBox-X has a "callout" system where the I/O and memory handlers are not populated directly but instead access to yet unknown regions triggers a search to any callout object with a bitmask matching that area. If the callout function returns a function, then that function is directly assigned and further I/O is direct and fast. It's up to the callout function to decide to respond or not. The purpose of this callout mechanism is to allow devices to easily emulate I/O port aliasing, or PnP devices that are relocatable.

By the way, don't forget that if you change the memory map... if the user is using specific cases of the dynamic core you need to call a function that forces dynamic core to follow memory map changes.

**Comment by joncampbell123** [comment]
PCI bus devices use base class PCI_Device declared in include/pci_bus.h. There are functions to register and unregister PCI device objects to specific bus and device slots. There is not yet code to support sub-devices. A good example would be the PCI VGA S3 chipset code and the 3Dfx Voodoo emulation.

**Comment by joncampbell123** [comment]
There is a timer callback system that is called once per 1ms of emulator time (probably related to the 1ms resolution offered by SDL_GetTicks). Your callback is called once every 1ms of emulator time, which might be sufficient for your purposes.

If you need events within the 1ms tick, there is PIC_AddEvent and PIC_RemoveEvents. All timeouts are in millseconds, NOT seconds, and they are floating point. They are long double on x86 and x86_64 using GCC, and double precision double for anything else including x86/x86_64 compiled with Microsoft Visual Studio.

**Comment by joncampbell123** [comment]
Nothing in terms of hardware emulation in DOSBox-X is multithreaded, same as any other DOSBox fork (although Staging and others say they're considering putting the 3Dfx software renderer in a separate thread).

**Comment by joncampbell123** [comment]
If you need to implement custom opcodes, the switch statement for opcodes in the normal core is in src/cpu/core_normal.cpp and src/cpu/core_normal/. Note the different headers for plain opcode vs those that start with 0Fh and whether or not the opcode is 32-bit or not. Note that modifying normal core will not put your custom opcodes in other cores like full core or the various dynamic cores.

**Comment by joncampbell123** [comment]
I think that should be enough to get you started, further details can be had by studying the source code. Have fun.

**Comment by joncampbell123** [comment]
For hardware interrupts, there are functions PIC_ActivteIRQ() and PIC_DeActivateIRQ().

If you mean setting up callbacks for software interrupts, look at the DOS kernel dos.cpp and the BIOS src/ints/bios.cpp to see how the interrupts are setup and handled.

**Comment by ThomasKjoernes** [comment]
Is there a fixed limit of 128 callbacks? Running the callbacks cmd in the debugger showed 75 as the last index used. Also index 0 and 4 had func value of 0 and blank desecription.

Would it be possible to "chain" callbacks to the same interrupts? For example if you wanted to extend INT21 handler, but you don't want to mess up dos.cpp. I guess I really want a "plugin" system :)

**Comment by joncampbell123** [comment]
Yes. The number of callbacks is limited by CB_MAX defined in include/callback.h.

If you want INT 21h plugins then perhaps instead of callback functions you should just add a function to register/unregister a list of functions to check before proceeding to normal INT 21h.

A good example of what you're talking about would be how DOSBox and DOSBox-X handle the INT 2Fh multiplex interrupt. There's a register/unregister API to extend what INT 2Fh can do.
